# Used cars on eBay
The data stems from 370,000 scraped from eBay Germany by Kaggle user orgesleka.

There has been a slight modification done to the dataset:
- A random sample of 50,000 was taken from the full dataset

|variable|description|
|:---|---:|
|dateCrawled | When this ad was first crawled. All field-values are taken from this date|
|name | Name of the car|
|seller | Whether the seller is private or a dealer|
|offerType | The type of listing|
|price | The price on the ad to sell the car|
|abtest | Whether the listing is included in an A/B test|
|vehicleType |The vehicle Type|
|yearOfRegistration | The year in which the car was first registered|
|gearbox | The transmission type|
|powerPS | The power of the car in PS|
|model | The car model name|
|kilometer |How many kilometers the car has driven|
|monthOfRegistration | The month in which the car was first registered|
|fuelType | What type of fuel the car uses|
|brand | The brand of the car|
|notRepairedDamage | If the car has a damage which is not yet repaired|
|dateCreated | The date on which the eBay listing was created|
|nrOfPictures | The number of pictures in the ad|
|postalCode | The postal code for the location of the vehicle|
|lastSeenOnline | When the crawler saw this ad last online|

## Objectives

## Conclusion

In [1]:
#Imports required
import pandas as pd
import numpy as np
import re as re
import datetime as dt

#Setting up dataset for use
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

There seems to be null values/missing entries in columns 6, 8, 10, 13 and 15 (vehicleType, gearbox, model, fuelType and notRepairedDamage).

However this may not pose much of an issue as it does not seem like a large portion of the dataset is null/missing.

Some of the variables could be represented as an int and float but are in reality in string format such as price and odometer.

Some date related variables could be converted to a datetime object.

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

There seems to be many spelling errors which need to be corrected as it may be confusing for other people who may want to use this dataset.

Several variable values seem to still be in german such as Angebot under offerType and nein under notRepairedDamage.

The variable names may need to be cleaned up as they seem to be somewhat not unified.

## Cleaning Columns
Now we are going to clean up the columns of the dataframe to make it more python friendly.

I looped through the column attribute which is an array containing all the columns. 

I used regex to substitute the strings which followed the pattern, a capital character followed by a small case character to be replaced by the output of the function lambda which replaced the original part of the string that matched the pattern and replaced it with an underscore along with the match case and applied `.group(0)` which is the original letter that is replaced.

Then I lower cased the output of the previous function which I reassigned back to `var_name` which is now fully clean.

I finally appended this to the clean list which I assigned back to the column attribute.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
#cleaning the column names
clean_list = []
for i in autos.columns:
    var_name = re.sub('[A-Z][a-z]',lambda x:"_"+x.group(0),i)
    '''Using regex sub function to find capital letters in the string and 
       replace it with a _ along with original letter'''
    var_name = var_name.lower()
    clean_list.append(var_name)
#assigning the list to become the new columns for the dataframe
autos.columns = clean_list

In [7]:
autos.describe()

year_of_registration       powerps  month_of_registration  \
count          50000.000000  50000.000000           50000.000000   
mean            2005.073280    116.355920               5.723360   
std              105.712813    209.216627               3.711984   
min             1000.000000      0.000000               0.000000   
25%             1999.000000     70.000000               3.000000   
50%             2003.000000    105.000000               6.000000   
75%             2008.000000    150.000000               9.000000   
max             9999.000000  17700.000000              12.000000   

       nr_of_pictures   postal_code  
count         50000.0  50000.000000  
mean              0.0  50813.627300  
std               0.0  25779.747957  
min               0.0   1067.000000  
25%               0.0  30451.000000  
50%               0.0  49577.000000  
75%               0.0  71540.000000  
max               0.0  99998.000000

It seems that `nr_of_pictures` column hold no meaningful data as it is mostly a full series with value of 0. 
This column can be dropped from the dataset.

The minimum value for `powerps` is 0 which is not realistic as cars require engine power to function. 
This needs to be investigated further.

The `year_of_registration` has some absurd minimum and maximum values (year of 1000 and 9999). 
This seems to imply these posts were both made in times cars did not exist or thousands of years in the future. This needs to be investigated.

`month_of_registration` seems to fine aside from the fact that the minimum value is 0 which is absurd as there is no 0 month.
This needs to be investigated.

There seems to be no issues with the `postal_code`. Upon a quick search into german postal codes, both the minimum and maximum postal codes are feasible.

In [8]:
autos['nr_of_pictures'].unique()

array([0])

This confirms the only input in `nr_of_pictures` is 0.

In [9]:
autos['powerps'].value_counts()

0        5500
75       3171
60       2195
150      2046
140      1884
         ... 
650         1
490         1
362         1
153         1
16312       1
Name: powerps, Length: 448, dtype: int64

There are 5500 entries with a powerps of 0.

Now we need to investigate the upper end of the spectrum as some of the engine ps recorded in the dataset is too large.

As a reference, the Bugatti Chiron which has been in production since 2016 can produce up to 1600ps worth of power.

In [10]:
autos.loc[autos['powerps']>1600,'powerps'].value_counts()

1800     3
1998     2
1753     1
15016    1
1986     1
1988     1
1796     1
16011    1
6226     1
8404     1
4400     1
15001    1
6045     1
2018     1
17700    1
3750     1
1704     1
2729     1
7511     1
1771     1
3500     1
6512     1
9011     1
1779     1
1780     1
1781     1
16312    1
14009    1
5867     1
1793     1
Name: powerps, dtype: int64

There is a possibility that these cars are privately upgraded and tuned however it is impossible to tell whether the powerps is true or not.

It might be best to drop these postings, which will not pose much of a problem as it is a very small sub-sample of the dataset. Another option is to replace these values with 'Unknown'.

In [11]:
before_1900 = autos['year_of_registration']<1900
after_2016_apr = (autos['year_of_registration']>2016)&(autos['month_of_registration']>3)
autos.loc[before_1900|after_2016_apr, ['name', 'brand', 'model','year_of_registration','price']]

name          brand  \
10                          VW_Golf_Tuning_in_siber/grau     volkswagen   
55                        Mercedes_E320_AMG_zu_Tauschen!  mercedes_benz   
65                         Ford_Fiesta_zum_ausschlachten           ford   
113                                Golf_4_Anfaenger_auto     volkswagen   
164    Opel_Meriva__nur_76000_Km__unfallfrei__scheckh...           opel   
...                                                  ...            ...   
49770                     VW_Polo_6n_Tuev_Neu!__1.6_75PS     volkswagen   
49796                        Opel_corsa_1.4_zu_verkaufen           opel   
49841                                  Passat_abzugeben.     volkswagen   
49880                                E39_528i_an_Bastler            bmw   
49935  Mercedes_A_klasse_angemeldet_mit_Tuef_und_Auto...  mercedes_benz   

          model  year_of_registration   price  
10          NaN                  2017    $999  
55     e_klasse                  2017      $1  
65       fiesta                  2017    $250  
113        golf                  2017  $1,200  
164      meriva                  2018  $4,800  
...         ...                   ...     ...  
49770      polo                  2018    $999  
49796     corsa                  2017  $4,500  
49841    passat                  2017    $600  
49880       5er                  2017      $0  
49935  a_klasse                  2017    $800  

[1108 rows x 5 columns]

In [12]:
autos.loc[before_1900|after_2016_apr,'year_of_registration'].value_counts()

2017    809
2018    286
2019      2
1800      2
5000      2
1500      1
1111      1
1001      1
1000      1
2800      1
4800      1
6200      1
Name: year_of_registration, dtype: int64

The year of registration for these cars are out of range for it to be accurate.

If we were to do analysis into registration date, these entries should be dropped or replaced with an Unknown value.

I personally would prefer to drop these as shown in the output above, it is most likely majority of the postings with out of range registration dates are adverts for car services or car swaps. 

In [13]:
autos['month_of_registration'].value_counts()

0     5075
3     5071
6     4368
5     4107
4     4102
7     3949
10    3651
12    3447
9     3389
11    3360
1     3282
8     3191
2     3008
Name: month_of_registration, dtype: int64

There are 5075 entries with a month of registration of 0 which does not make sense.

In [14]:
autos.loc[autos['month_of_registration']==0,['brand', 'model','month_of_registration', 'year_of_registration','price']]

brand   model  month_of_registration  year_of_registration  \
11      mercedes_benz     NaN                      0                  2000   
18                bmw     3er                      0                  1995   
27               ford     NaN                      0                  2005   
46                bmw     NaN                      0                  2005   
52               opel  andere                      0                  1985   
...               ...     ...                    ...                   ...   
49932   mercedes_benz     NaN                      0                  1999   
49937            audi      a6                      0                  1996   
49984  sonstige_autos     NaN                      0                  2000   
49985      volkswagen     NaN                      0                  1995   
49991         renault  twingo                      0                  2016   

        price  
11       $350  
18       $300  
27         $0  
46     $9,000  
52     $3,500  
...       ...  
49932  $1,400  
49937    $999  
49984      $0  
49985  $1,000  
49991    $500  

[5075 rows x 5 columns]

These most likely would need to be replaced with Unknown value as it is difficult to find out which month these cars were registered. 

The same car model can generally registered over a range of years at which it is produced.

I think it is better to replace these months with 0 as unknown as unlike year of registration (shown on the number plate) month is more difficult to find hence why sellers are more likely just to place it as 0 as they do not know what month it was registered.

In [15]:
autos.describe(include = 'object')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-14 20:50:02  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   

       vehicle_type  gearbox  model   odometer fuel_type       brand  \
count         44905    47320  47242      50000     45518       50000   
unique            8        2    245         13         7          40   
top       limousine  manuell   golf  150,000km    benzin  volkswagen   
freq          12859    36993   4024      32424     30107       10687   

       not_repaired_damage         date_created            last_seen  
count                40171                50000                50000  
unique                   2                   76                39481  
top                   nein  2016-04-03 00:00:00  2016-04-07 06:17:27  
freq                 35232                 1946                    8

As stated before, price and odometer seem to be able to be represented as an int/float object.

Price seems to suffer from the same issues as the int objects above as 0 seems to be a popular price. This is most likely due to the fact sellers put 0 as price to attract buyers and state the price in the description or after the buyer has contacted the seller. The same would probably be true with very low values for price e.g. 1, 5, 10. However we should be careful as to if `no_repaired_damage` is set to true.

Some variables seem to have missing data such as vehicle_type, gearbox, model etc.

In [16]:
#Before looking at price and odometer further, we need to convert to int
autos['price'].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [17]:
#Need to remove dollar sign and commas
autos['price'] = autos['price'].str.replace(',','')
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].astype(int)

In [18]:
autos['odometer'].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [19]:
#Need to remove comma and km
autos['odometer'] = autos['odometer'].str.replace(',','')
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].astype(int)

In [20]:
#Now we need to rename variable names so we know the units for later use
autos.rename({'price':'price_usd'}, axis=1, inplace=True)
autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price_usd   abtest vehicle_type  year_of_registration  \
0  privat    Angebot       5000  control          bus                  2004   
1  privat    Angebot       8500  control    limousine                  1997   
2  privat    Angebot       8990     test    limousine                  2009   
3  privat    Angebot       4350  control   kleinwagen                  2007   
4  privat    Angebot       1350     test        kombi                  2003   

     gearbox  powerps   model  odometer_km  month_of_registration fuel_type  \
0    manuell      158  andere       150000                      3       lpg   
1  automatik      286     7er       150000                      6    benzin   
2    manuell      102    golf        70000                      7    benzin   
3  automatik       71  fortwo        70000                      6    benzin   
4    manuell        0   focus       150000                      7    benzin   

        brand not_repaired_damage         date_created  nr_of_pictures  \
0     peugeot                nein  2016-03-26 00:00:00               0   
1         bmw                nein  2016-04-04 00:00:00               0   
2  volkswagen                nein  2016-03-26 00:00:00               0   
3       smart                nein  2016-03-12 00:00:00               0   
4        ford                nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [21]:
autos['price_usd'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

There seems to be a mean price of 9840USD and a minimum of 0USD which is not feasible as realistically nobody will give away a vehicle away for free. There is a maximum price of 10,000,000. This may be possible for very high end/rare cars however may need to be inspected.

In [22]:
autos.loc[autos['price_usd']>350000,['brand','model','price_usd','year_of_registration']]

brand        model  price_usd  year_of_registration
514              ford        focus     999999                  2009
2897             ford       escort   11111111                  1973
7814   sonstige_autos          NaN    1300000                  1992
11137  sonstige_autos          NaN   10000000                  1960
22947             bmw          NaN    1234566                  1999
24384      volkswagen          NaN   11111111                  1995
27371            fiat        punto   12345678                  2017
37585      volkswagen        jetta     999990                  1985
39377           volvo          v40   12345678                  2018
39705   mercedes_benz     s_klasse   99999999                  1999
42221         citroen           c4   27322222                  2014
43049      volkswagen  transporter     999999                  1981
47598            opel       vectra   12345678                  2001
47634  sonstige_autos          NaN    3890000                  2006

These cars are very common cars and should not be coming with the price tags. These must be listings with random price tags with the actual price labelled in the description. 

In [23]:
less_50 = autos['price_usd']<50
not_0 = autos['price_usd']!=0
autos.loc[less_50&not_0,['brand','model','price_usd','year_of_registration']]

brand     model  price_usd  year_of_registration
55     mercedes_benz  e_klasse          1                  2017
64               bmw       5er         40                  2011
344            honda    andere          1                  2000
791             opel    vectra          1                  1999
981              bmw       3er          1                  2003
...              ...       ...        ...                   ...
49240            bmw       NaN          1                  1999
49377            bmw       3er          1                  2002
49528            bmw       3er          1                  2016
49728           ford     focus          1                  2016
49845     volkswagen    sharan          1                  2000

[218 rows x 4 columns]

As predicted there are many listings with just a price of 1 or a really low price which is most likely to attract buyers to click on the advert.

I want to drop all the entries where the price is below 10.

In [24]:
autos_2 = autos.loc[(autos['price_usd']>10)&(autos['price_usd']<350000),:].copy()

In [25]:
autos_2['odometer_km'].describe()

count     48393.000000
mean     125807.451491
std       39710.298822
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

Eyeballing the describe statistics of odometer_km there seems to be no obvious outliers so data does not need to be inspected.

In [26]:
autos_2['date_crawled'] = pd.to_datetime(autos_2['date_crawled'])
autos_2['last_seen'] = pd.to_datetime(autos_2['last_seen'])
autos_2['date_created'] = pd.to_datetime(autos_2['date_created'])

In [27]:
autos_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48393 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   date_crawled           48393 non-null  datetime64[ns]
 1   name                   48393 non-null  object        
 2   seller                 48393 non-null  object        
 3   offer_type             48393 non-null  object        
 4   price_usd              48393 non-null  int64         
 5   abtest                 48393 non-null  object        
 6   vehicle_type           43883 non-null  object        
 7   year_of_registration   48393 non-null  int64         
 8   gearbox                46105 non-null  object        
 9   powerps                48393 non-null  int64         
 10  model                  45965 non-null  object        
 11  odometer_km            48393 non-null  int64         
 12  month_of_registration  48393 non-null  int64         
 13  f

In [28]:
autos_2['date_crawled'].dt.strftime('%Y-%m-%d').value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025376
2016-03-06    0.014052
2016-03-07    0.036038
2016-03-08    0.033269
2016-03-09    0.033042
2016-03-10    0.032215
2016-03-11    0.032608
2016-03-12    0.036968
2016-03-13    0.015643
2016-03-14    0.036638
2016-03-15    0.034282
2016-03-16    0.029508
2016-03-17    0.031595
2016-03-18    0.012915
2016-03-19    0.034757
2016-03-20    0.037795
2016-03-21    0.037299
2016-03-22    0.032939
2016-03-23    0.032277
2016-03-24    0.029405
2016-03-25    0.031554
2016-03-26    0.032257
2016-03-27    0.031079
2016-03-28    0.034860
2016-03-29    0.034117
2016-03-30    0.033745
2016-03-31    0.031823
2016-04-01    0.033745
2016-04-02    0.035501
2016-04-03    0.038580
2016-04-04    0.036493
2016-04-05    0.013080
2016-04-06    0.003162
2016-04-07    0.001384
Name: date_crawled, dtype: float64

There seems to be no obvious underlying pattern other than the fact that this dataset was collected in March and April 2016.

Using this I have updated the above code about year and month of registration and what is the feasible ranges.

In [29]:
autos_2['date_created'].dt.strftime('%Y-%m-%d').value_counts(normalize=True, dropna=False)

2016-04-03    0.038828
2016-03-20    0.037857
2016-03-21    0.037526
2016-04-04    0.036865
2016-03-12    0.036782
                ...   
2016-02-01    0.000021
2016-02-22    0.000021
2015-09-09    0.000021
2016-01-14    0.000021
2015-11-10    0.000021
Name: date_created, Length: 76, dtype: float64

This seems to be very disaggregated. 
Let's have a look at a more broader view by month.

In [30]:
autos_2['date_created'].dt.strftime('%Y-%m').value_counts(normalize=True, dropna=False)

2016-03    0.837476
2016-04    0.160891
2016-02    0.001261
2016-01    0.000248
2015-12    0.000041
2015-09    0.000021
2015-08    0.000021
2015-11    0.000021
2015-06    0.000021
Name: date_created, dtype: float64

The first ever posting in this dataset was on the date 11th June 2015 whilst the latest posting is on the 7th April 2016. 

Looking at the posts by month, it seems majority of the posting was posted within the month the post was crawled from ebay. This implies that majority of the cars tend to sell fairly quickly (within 3 months).

In [31]:
#Now I want to clean up the data with erraneous year and month of registration
autos_3 = autos_2.loc[(autos['year_of_registration']>1900)&(autos['year_of_registration']<2016),:].copy()

In [32]:
autos_3.shape

(45310, 20)

In [33]:
#Change month = 0 to null
autos_3.loc[autos_3['month_of_registration']==0,'month_of_registration'] = np.nan

In [34]:
autos_3.shape

(45310, 20)

What I have done to the dataset so far:

Removed listings where price was less than 10USD and above 350,000USD.

Removed all listings where year of registration is below 1900s and later than 2016. 

Changed value of month of registration to null where month of registration is listed as 0.

I will not make changes to engine ps as I will not be conducting analysis on it.

In [35]:
#Average prices split up by brand
mean_dict = {}
for i in autos_3['brand'].unique():
    average = autos_3.loc[autos_3['brand']==i,'price_usd'].mean()
    mean_dict[i] = average

In [36]:
list_1 = list(mean_dict.items())
mean_list = []
for i in list_1:
    value = i[1]
    key = i[0]
    mean_list.append([value, key])

In [37]:
mean_list.sort(reverse=True)
mean_list

[[44130.43165467626, 'porsche'],
 [18585.447916666668, 'land_rover'],
 [12060.013667425968, 'sonstige_autos'],
 [11834.163461538461, 'jeep'],
 [11635.493150684932, 'jaguar'],
 [10653.556390977443, 'mini'],
 [9373.358442871588, 'audi'],
 [8739.202448424394, 'mercedes_benz'],
 [8431.807310290027, 'bmw'],
 [6743.984674329502, 'chevrolet'],
 [6366.592838196287, 'skoda'],
 [6015.504587155963, 'kia'],
 [5986.446280991736, 'dacia'],
 [5495.287375589313, 'volkswagen'],
 [5466.280087527352, 'hyundai'],
 [5231.613793103448, 'toyota'],
 [5019.739234449761, 'volvo'],
 [4828.492774566474, 'nissan'],
 [4445.214896214896, 'seat'],
 [4190.1511627906975, 'mazda'],
 [4178.645070422535, 'honda'],
 [4163.850993377483, 'alfa_romeo'],
 [4152.630597014925, 'suzuki'],
 [4044.618556701031, 'subaru'],
 [3834.635220125786, 'citroen'],
 [3825.544908285895, 'ford'],
 [3616.225505443235, 'smart'],
 [3514.1739130434785, 'chrysler'],
 [3490.1054054054052, 'mitsubishi'],
 [3376.22, 'lancia'],
 [3243.684210526316, 'saa

On the eBay marketplace it seems porsche's on average are listed for the largest price. It doubles the average price for a land rover.  This is probably due to the face porche only making high performance sport vehicles, whilst land rover does produce luxury SUVs, they are generally sold at a much lower RRP compared to a brand new porsche. 

What I find surprising is that mini coopers are on average more costly in the german market compared to higher tier german vehicles such as bmw, audi and mercedes. This is most likely due to the popularity of german vehicles in germany thus there is an abundance in these brand vehicles however the opposite could be said for the british vehicles land rover, jaguar and mini. 

Sonstige_autos is the third most expensive option on eBay Germany. This other category is most likely for cars which have modifications done to them for example Brabus or Volvo's polestar etc. Since these cars would have expensive additions to them, they would most likely cost more than a stock version of the cars itself.

In [38]:
#Average mileage split up by brand
mean_mileage_dict = {}
for i in autos_3['brand'].unique():
    average_km = autos_3.loc[autos_3['brand']==i,'odometer_km'].mean()
    mean_mileage_dict[i] = average_km

In [47]:
#Convert dictionaries to series using the constructor
bmp_series = pd.Series(mean_dict, dtype='float64')
bmo_series = pd.Series(mean_mileage_dict, dtype='float64')

brand_averages = pd.DataFrame(bmp_series, columns = ['mean_price'])
brand_averages['mean_mileage'] = bmo_series

In [59]:
brand_averages.sort_values(by='mean_price', axis = 0, ascending = False)

mean_price   mean_mileage
porsche         44130.431655   97859.712230
land_rover      18585.447917  120208.333333
sonstige_autos  12060.013667   90956.719818
jeep            11834.163462  126682.692308
jaguar          11635.493151  124178.082192
mini            10653.556391   88433.583960
audi             9373.358443  129246.713852
mercedes_benz    8739.202448  130671.049649
bmw              8431.807310  132446.364720
chevrolet        6743.984674   99865.900383
skoda            6366.592838  110901.856764
kia              6015.504587  112813.455657
dacia            5986.446281   84586.776860
volkswagen       5495.287376  128449.449974
hyundai          5466.280088  105754.923414
toyota           5231.613793  115568.965517
volvo            5019.739234  138098.086124
nissan           4828.492775  118410.404624
seat             4445.214896  121007.326007
mazda            4190.151163  124004.360465
honda            4178.645070  122267.605634
alfa_romeo       4163.850993  131490.066225
suzuki           4152.630597  108675.373134
subaru           4044.618557  126804.123711
citroen          3834.635220  119544.025157
ford             3825.544908  123989.563567
smart            3616.225505   98685.847589
chrysler         3514.173913  132049.689441
mitsubishi       3490.105405  126351.351351
lancia           3376.220000  121900.000000
saab             3243.684211  144342.105263
peugeot          3137.947133  127081.161579
opel             3031.272917  129261.458333
fiat             2866.012163  116589.921807
lada             2688.296296   83518.518519
renault          2500.814130  128013.276434
trabant          1790.861538   54538.461538
daihatsu         1654.478261  116043.478261
rover            1644.866667  138416.666667
daewoo           1076.764706  120808.823529

The mileage sheds some light into why the german brands on average got listed for less than a mini. The german brands on average got listed with an average mileage of around 30000 higher. 

The average mileage of a used car is fairly high across the board. 